pkg/sql/lease_internal_test.go

```diff
@@ -558,3 +562,77 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
 
 	wg.Wait()
 }
+
+// This test makes sure the lease gets renewed automatically in the background.
+func TestLeaseRefreshedAutomatically(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	var testAcquiredCount int32
+	testingKnobs := base.TestingKnobs{
+		SQLLeaseManager: &LeaseManagerTestingKnobs{
+			LeaseStoreTestingKnobs: LeaseStoreTestingKnobs{
+				// We want to track what leases get acquired,
+				LeaseAcquiredEvent: func(table sqlbase.TableDescriptor, _ error) {
+					if table.Name == "test" {
+						atomic.AddInt32(&testAcquiredCount, 1)
+					}
+				},
+			},
+		},
+	}
+	s, sqlDB, kvDB := serverutils.StartServer(t, base.TestServerArgs{Knobs: testingKnobs})
+	defer s.Stopper().Stop(context.TODO())
+	leaseManager := s.LeaseManager().(*LeaseManager)
+
+	if _, err := sqlDB.Exec(`
+CREATE DATABASE t;
+CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR);
+`); err != nil {
+		t.Fatal(err)
+	}
+	now := s.Clock().Now()
+	tableDesc := sqlbase.GetTableDescriptor(kvDB, "t", "test")
+
+	savedLeaseRenewalTimeout := LeaseRenewalTimeout
+	defer func() {
+		LeaseRenewalTimeout = savedLeaseRenewalTimeout
+	}()
+	LeaseRenewalTimeout = 10 * time.Millisecond
+
+	// Acquire the first lease. This begins the refreshing routine.
+	_, e1, err := leaseManager.Acquire(context.TODO(), now, tableDesc.ID)
+	if err != nil {
+		t.Error(err)
+	}
+
+	// Set the lease duration to be higher so that the new refreshed lease will
+	// be guaranteed to have a higher expiry time, otherwise jitter may cause it
+	// to have a lower expiry time than the previous lease. We also need to set
+	// the LeaseRenewalTimeout to be high so we renew immediately.
+	savedLeaseDuration := LeaseDuration
+	defer func() {
+		LeaseDuration = savedLeaseDuration
+	}()
+	LeaseDuration = 2 * LeaseDuration
+
+	// Keep checking for a new lease to be acquired by the refreshing routine.
+	testutils.SucceedsSoon(t, func() error {
+		// Acquire the newer lease.
+		ts, e2, err := leaseManager.Acquire(context.TODO(), now, tableDesc.ID)
+		if err != nil {
+			t.Error(err)
+		}
+
```

Comment:
move the defer here so that a successful Acquire() always results in a Release()

Reply:
This is done intentionally because if we release the first lease the call to acquire will block to get a new lease as opposed to the refresh routine acquiring it.

I'll add a comment to explicitly call this out.

Reply:
I don't think that's the case. Releasing a lease doesn't result in a new lease being acquired on the next acquire(). The refcount on a lease can go down to 0 without it resulting in it being released.